internal/core/subvolume.cpp

```diff
@@ -0,0 +1,108 @@
+#include <stdexcept>
+
+#include <boost/math/interpolators/makima.hpp>
+
+#include "axis.hpp"
+#include "subvolume.hpp"
+
+SubVolume* make_subvolume(
+    DataHandle& handle,
+    RegularSurface const& reference,
+    RegularSurface const& top,
+    RegularSurface const& bottom
+) {
+    if (!(reference.plane() == top.plane() && reference.plane() == bottom.plane())) {
+        throw std::runtime_error("Expected surfaces to have the same plane and size");
+    }
+
+    MetadataHandle const& metadata = handle.get_metadata();
+    auto transform = metadata.coordinate_transformer();
+
+    auto iline = metadata.iline();
+    auto xline = metadata.xline();
+    auto sample = metadata.sample();
+
+    RawSegmentBlueprint segment_blueprint = RawSegmentBlueprint(sample.stepsize(), sample.min());
+    auto subvolume = new SubVolume(reference, top, bottom, segment_blueprint);
+    auto const horizontal = subvolume->horizontal();
+
+    subvolume->m_segment_offsets[0] = 0;
+    for (int i = 0; i < horizontal.size(); ++i) {
```

Comment:
I introduced bounded plane last time to **avoid** having to write reference.size here :joy_cat: 
It was too confusing to me - why do we have to iterate over reference surface? Why not the top? To me it makes more sense to say that I want to iterate over horizontal surface of subvolume.

It might be nicer to get reference_depth, top_dept and bottom_depth like subvolume.reference(i), subvolume.top(i), subvolume.bottom(i), but as I was fine with current code I didn't add those. Can do, if it helps you.

I originally wanted to hide all these calculations in subvolume, but it is still kinda a creation of it, so adding methods didn't make much sense.

Reply:
Not a big deal, but I have some problems following the argument here. Horizontal is a new term for reference.plane.

SubVolume is a class containing data and related code. For me it is natural that the class is responsible for consistency defined by the underlaying object. Currently it is not clear to me how is responsible for ensuring that top, ref and bottom surface is of the same size. Using reference.size directly would limit this from two to three checks.